### PR TITLE
Force the buffer to be flushed during bootstrapping

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington 

Also refactored bootstrapping logic into a separate `bootstrapManager` interface.

### CODE REVIEW

File: storage/bootstrap.go [hunk 1]
```diff
@@ -33,37 +36,111 @@ const (
...
+type bootstrapManager interface {
+	// IsBootstrapped returns whether the database is already bootstrapped.
+	IsBootstrapped() bool
+
+	// Bootstrap performs bootstrapping for all shards owned by db. It returns an error
+	// if the server is currently being bootstrapped, and nil otherwise.
+	Bootstrap() error
+}
+
+type bsManager struct {
```
Comment (robskillington) [inline storage/bootstrap.go]: To follow conventions of rest of `storage` you could call the interface `databaseBootstrapManager` and the struct `bootstrapManager` perhaps?

File: storage/bootstrap.go [hunk 2]
```diff
@@ -33,37 +36,111 @@ const (
...
+	sync.RWMutex
+
+	database *db                  // storage database
```
Comment (robskillington) [inline storage/bootstrap.go]: Can we avoid relying on the concrete `db` struct and rely on the interface itself?  This way we can test the component in complete isolation and always the dependencies, which is the pattern I've been trying to keep amongst all the code so far.

Comment (robskillington) [inline storage/bootstrap.go]: I see so right now the `Database` interface is expored, so perhaps we can create another interface `database` that is not export that extends `Database`?

e.g.,
```go
type database interface {
  Database

  //.. methods
}
```

File: storage/buffer.go [hunk 3]
```diff
@@ -161,12 +161,19 @@ func (s *dbBuffer) bucketState(
...
+		needsDrain, needsReset := true, true
+		if !forced {
+			_, needsDrain, needsReset = s.bucketState(now, b, current)
+			if !needsDrain && !needsReset {
+				// No action necessary
+				return
+			}
+		} else if b.drained || b.lastWriteAt.IsZero() {
+			// No action necessary if the bucket has been drained,
+			// or there is no data in the bucket.
```
Comment (robskillington) [inline storage/buffer.go]: Might be more clean like so:

```go
		_, needsDrain, needsReset := s.bucketState(now, b, current)
		if forced && !b.drained && !b.lastWriteAt.IsZero() {
			// Forcing drain and reset, this block hasn't been drained and has data
			needsDrain, needsReset = true, true
		}
		if !needsDrain && !needsReset {
			// No action necessary
			return
		}
```

File: storage/series.go [hunk 4]
```diff
@@ -283,39 +283,46 @@ func (s *dbSeries) drainStream(blocks m3db.DatabaseSeriesBlocks, stream io.Reade
...
-func (s *dbSeries) Bootstrap(rs m3db.DatabaseSeriesBlocks, cutover time.Time) error {
-	if success, err := tryBootstrap(&s.RWMutex, &s.bs, "series"); !success {
-		return err
-	}
+// NB(xichen): we are holding a big lock here to drain the in-memory buffer.
+// This could potentially be expensive in that we might accumulate a lot of
+// data in memory during bootstrapping. If that becomes a problem, we could
+// bootstrap in batches, e.g., drain and reset the buffer, drain the streams,
+// then repeat, until len(s.pendingBootstrap) is below a given threshold.
+func (s *dbSeries) Bootstrap(rs m3db.DatabaseSeriesBlocks, cutOver time.Time) error {
```
Comment (robskillington) [inline storage/series.go]: Why the change from `cutOver` to `cutover` out of interest?  Cutover seems to be a single word: 
http://www.dictionary.com/browse/cutover

File: storage/bootstrap.go (hunk 1 above)
Comment (xichen2020) [inline storage/bootstrap.go]: yeah sounds good

File: storage/bootstrap.go (hunk 2 above)
Comment (xichen2020) [inline storage/bootstrap.go]: yea can do

File: storage/series.go (hunk 4 above)
Comment (xichen2020) [inline storage/series.go]: lol thought those were two words, will change it back

File: storage/buffer.go (hunk 3 above)
Comment (xichen2020) [inline storage/buffer.go]: yeah i guess it doesn't hurt to drain/reset when it's forced and the first if condition is not satisfied. Will update.

File: storage/bootstrap.go [hunk 5]
```diff
@@ -33,37 +36,111 @@ const (
...
+
+type bootstrapManager struct {
+	sync.RWMutex
+
+	database database             // storage database
+	opts     m3db.DatabaseOptions // database options
+	state    bootstrapState       // bootstrap state
+}
+
+func newBootstrapManager(database *db) databaseBootstrapManager {
```
Comment (robskillington) [inline storage/bootstrap.go]: Do you want to make the constructor take the interface rather than the struct pointer?

Comment (robskillington) [inline storage/bootstrap.go]: Make this take the interface instead of struct pointer?

Comment (robskillington) [inline storage/bootstrap.go]: Make this take the interface instead of struct pointer?

Comment (xichen2020) [inline storage/bootstrap.go]: Of course, updated

**Comment by robskillington** [comment]
LGTM :shipit:
